core/Pos/Binary/Core/Coin.hs

```diff
@@ -115,20 +120,25 @@ hdrToParam h
     | isClear h 4 = (3, fromIntegral (h .&. 0x0f))
     | otherwise   = (4, fromIntegral (h .&. 0x0f))
 
-decodeVarint :: Peek Word64
-decodeVarint = do
-    (nbBytes, acc) <- hdrToParam <$> getWord8
-    conts <- replicateM nbBytes getWord8
-    let val = orAndShift acc conts
-    when (expectedContBytes val /= length conts) $ fail "not canonical encoding"
-    return val
+decodeMegaMicros :: [Word8] -> Decoder s (Word64, [Word8])
```

Comment:
Where possible I'd like to use standard CBOR encodings. I don't yet fully understand the coin encoding issue however.